### ISSUE TITLE
[Feature Request] Asynchronous run of "Scanning For Items" and "Scanning Item" tasks

### ISSUE THREAD

**Issue by Searinox** [mgba-emu/mgba#1952]
On 3DS with about 2000+ ROMs in folder it is slow to complete the "Scanning For Items" phase but after that is done there is an additional "Scanning Item" task after that finishes. The "Scanning Item" task completes much faster.

If there would be one thread for each of these tasks and the "Scanning Item" task would asynchronously be scanning the ROMs as the "Scanning For Items" task loads them in, there would be next to no extra waiting time after "Scanning For Items" because the "Scanning Item" task will have already processed them.

Since "Scanning Item" is much faster, it will always be caught up with the new ROMs being loaded in by "Scanning For Items", leaving it as the only bottleneck instead of running both tasks sequentially.

**Comment by endrift** [comment]
In theory, what these loops are supposed to do is:
- The first loop, scanning for items, is just a directory traversal. It gathers a list of files and folders in the directory.
- The second loop, scanning items, figures out which items should and shouldn't be displayed.

The problem is that both loops are I/O heavy, which does not thread well. Even if it were done asynchronously it wouldn't solve the I/O bottleneck.

The thing is, it turned out that opening each and every file to check if it was a ROM or not turned out to be ungodly slow, so I replaced it with a filename filter that just filters out savestates and the like.

BUT! The filename filter is FAST. There's no reason the second loop is even needed at that point, so I've pushed a change that moves the filename filter to the first loop and disables the second loop (or at least, makes it quiet--it's still needed to pick the right filename to pre-select since the first loop outputs an unsorted list, and we need an index into the sorted list, but that's very fast too) if you're just filtering filenames.

Further, I found a way to speed up listing directories on 3DS so I just pushed that too.
